keystrokes/src/components/Game/Game.js

```diff
@@ -1,98 +1,112 @@
-
-			this.setState(
-				{
-					keybindings,
-					keys: Object.keys(keybindings),
-					key: '1',
-				},
-				() => {
-					const newKey = this.getNextKey(this.state.keys);
-					this.setState({ key: newKey });
-				}
-			);
-		} catch (error) {
-			console.warn(error);
+				collectKeybindings(
+					charDetails,
+					charDetails.selectedSpec,
+					charDetails.specs[charDetails.selectedSpec]
+						.selectedKeybindings
+				);
+			}
 		}
-	}
+		async function collectKeybindings(character, spec, keyBinding) {
+			const path = `/Keybindings/${characterKeybindings(
+				character,
+				spec,
+				keyBinding
+			)}`;
+			const snapShot = await ref.child(path).once('value');
+			if (snapShot.exists()) {
+				const k = snapShot.val();
+				setKeyBindings(k);
+				const newKey = getNextKey(Object.keys(k));
+				setKey(newKey);
+			}
+		}
+		collectCharacterInfo();
+	}, [userInfo]);
+
+	useEffect(() => {
+		document.body.onkeydown = handleKeyPress;
+		return () => {
+			document.body.onkeydown = null;
+		};
+	}, [keyBindings, key]);
 
-	handleKeyPress = e => {
+	function handleKeyPress(e) {
 		if (!e.metaKey) {
 			e.preventDefault();
 		}
-		console.log(e.code);
 		const keyPressed = {
-			key: e.code.toLowerCase().replace(/digit|key|left|right/i, ''),
-			altKey: e.altKey,
-			ctrlKey: e.ctrlKey,
-			shiftKey: e.shiftKey,
+			key: verifyKey(e.code.toLowerCase().replace(/digit|key/i, '')),
+			Alt: e.altKey,
+			Ctrl: e.ctrlKey,
+			Shift: e.shiftKey,
+			None: !(e.altKey || e.ctrlKey || e.shiftKey),
 		};
-		if (
-			keyPressed.key !== 'shift' &&
-			keyPressed.key !== 'alt' &&
-			keyPressed.key !== 'control'
-		) {
-			const expectedKey = this.state.keybindings[this.state.key];
-			//there is an issue here if the key is not a letter or a number, such as ` or ,
-			//also issue with tab related commands (ctrl w)
+		if (validatePress(keyPressed.key)) {
+			const expectedKey = keyBindings[key];
 			if (
-				keyPressed.key === expectedKey.key &&
-				((expectedKey.modifier === 'CONTROL' && keyPressed.ctrlKey) ||
-					(expectedKey.modifier === 'SHIFT' && keyPressed.shiftKey) ||
-					(expectedKey.modifier === 'ALT' && keyPressed.altKey) ||
-					(expectedKey.modifier === 'NONE' &&
-						!keyPressed.ctrlKey &&
-						!keyPressed.altKey &&
-						!keyPressed.shiftKey))
+				keyPressed.key === expectedKey.Key &&
+				keyPressed[expectedKey.Mod]
 			) {
-				const newKey = this.getNextKey(this);
-				this.setState({ key: newKey });
+				const newKey = getNextKey(Object.keys(keyBindings));
+				setKey(newKey);
+				setFailedFirstTry(false);
+			} else {
+				setFailedFirstTry(true);
 			}
 		}
-	};
-
-	render() {
-		return (
-			<Fragment>
-				<div className="App-header">
-					{this.state.keybindings[this.state.key].spell}
-				</div>
-				<div className="App-header">
-					On {this.state.keybindings[this.state.key].target}
-				</div>
-			</Fragment>
-		);
 	}
+
+	return (
+		<React.Fragment>
```

Comment:
I tried this with <> </> instead of a Fragment, but it didnt work. Any idea why?

Reply:
Not sure without an error,  no big deal